### PR TITLE
Move docker system information to a dedicated router and backend.

### PR DESCRIPTION
**Pull request by calavera**
Because I like the name `system` better than `local` :)

This leave us only with the image backend to refactor. But we need to wait until #18347 is merged before doing that.

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

**Comment by MHBauer** [comment]
Agree on waiting for #18347. Agree on this split of 'systemy' functions into separate router. I can see what's left is 'imagey' functions + build.

File: daemon/info.go [hunk 1]
```diff
@@ -113,6 +113,25 @@ func (daemon *Daemon) SystemInfo() (*types.Info, error) {
 	return v, nil
 }
 
+// SystemVersion returns version information about the daemon.
+func (daemon *Daemon) SystemVersion() types.Version {
```
Comment (MHBauer) [inline daemon/info.go]: Why move this into the daemon if it can be handled outside? I think the function itself looks fine.

Comment (calavera) [inline daemon/info.go]: this allows backends to change the information in the version. I don't know if it's recommendable, but I'm just following conventions :smile_cat:

Comment (MHBauer) [inline daemon/info.go]: Oh, that's good. Duh doi, on my part. I agree with this line of thought. :+1:

File: api/server/router/system/backend.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+)
+
+// Backend is the methods that need to be implemented to provide
+// system specific functionality.
+type Backend interface {
+	SystemInfo() (*types.Info, error)
+	SystemVersion() types.Version
+	SubscribeToEvents(since, sinceNano int64, ef filters.Args) ([]*jsonmessage.JSONMessage, chan interface{})
+	UnsubscribeFromEvents(chan interface{})
+	AuthenticateToRegistry(authConfig *cliconfig.AuthConfig) (string, error)
```
Comment (runcom) [inline api/server/router/system/backend.go]: maybe I might have missed something but seems like backends would be pluggable? (just asking, I love this interface here)

**Comment by runcom** [comment]
LGTM! so clean (just a curiosity)

File: api/server/router/system/backend.go (hunk 2 above)
Comment (calavera) [inline api/server/router/system/backend.go]: Yes, that's the idea, I still don't know how completely true it's going to be, but we can always iterate.

For instance, a first use case is that swarm should be able to reuse the api endpoints and write their own backends.

Comment (runcom) [inline api/server/router/system/backend.go]: amazing, ok, thanks :)

**Comment by estesp** [comment]
LGTM

File: api/server/router/system/backend.go (hunk 2 above)
Comment (MHBauer) [inline api/server/router/system/backend.go]: Isn't it great! :smiling_imp:

**Pull request merged**